backends/utils/history_server.go

```diff
@@ -0,0 +1,363 @@
+package utils
+
+import (
+	"bufio"
+	"bytes"
+	"encoding/json"
+	"fmt"
+	"os"
+	"path"
+	"time"
+
+	"github.com/google/uuid"
+	"github.com/nuclio/logger"
+	"github.com/pkg/errors"
+	"github.com/v3io/frames"
+	"github.com/v3io/frames/v3ioutils"
+	v3io "github.com/v3io/v3io-go/pkg/dataplane"
+	v3iohttp "github.com/v3io/v3io-go/pkg/dataplane/http"
+	"github.com/v3io/v3io-tsdb/pkg/utils"
+)
+
+const (
+	writeMonitoringLogsTimeout = 10 * time.Second
+	pendingLogsBatchSize       = 100
+	logsPathPrefix             = "/monitoring/frames"
+	logsContainer              = "users"
+	maxBytesInNginxRequest     = 5 * 1024 * 1024 // 5MB
+
+	queryType = "query"
+)
+
+type HistoryEntry struct {
+	ActionType     string
+	UserName       string
+	BackendName    string
+	TableName      string
+	Container      string
+	ActionDuration time.Duration
+	StartTime      time.Time
+
+	AdditionalData map[string]string
+}
+
+type HistoryServer struct {
+	logFilePath string
+	logger      logger.Logger
+	logs        []HistoryEntry
+	requests    chan HistoryEntry
+	container   v3io.Container
+	config      *frames.Config
+	isActive    bool
+}
+
+func NewHistoryServer(logger logger.Logger, cfg *frames.Config) *HistoryServer {
+	logPath := fmt.Sprintf("%v/%v_%v.json", logsPathPrefix, uuid.New().String(), time.Now().Unix())
+
+	mon := HistoryServer{logFilePath: logPath,
+		logger:   logger,
+		requests: make(chan HistoryEntry, 100),
+		isActive: false,
+		config:   cfg}
+
+	mon.createDefaultV3ioClient()
+	if mon.isActive {
+		mon.Start()
+	}
+	return &mon
+}
+
+func (m *HistoryServer) Start() {
+	go func() {
+		var pendingLogs []HistoryEntry
+
+		for {
+			select {
+			case entry := <-m.requests:
+				pendingLogs = append(pendingLogs, entry)
+				m.logs = append(m.logs, entry)
+
+				if len(pendingLogs) == pendingLogsBatchSize {
+					m.writeMonitoringBatch(pendingLogs)
+					pendingLogs = pendingLogs[:0]
+				}
+			case <-time.After(writeMonitoringLogsTimeout):
+				if len(pendingLogs) > 0 {
+					m.writeMonitoringBatch(pendingLogs)
+					pendingLogs = pendingLogs[:0]
+				}
+			}
+		}
+	}()
+}
+
+func (m *HistoryServer) createDefaultV3ioClient() {
+
+	session := &frames.Session{}
+
+	token := os.Getenv("V3IO_ACCESS_KEY") // "0938b76f-79ff-41f6-981d-864ea3291cca"
+	if token == "" {
+		m.logger.Warn("can not create v3io.client. could not find `V3IO_ACCESS_KEY` environment variable")
+		return
+	}
+
+	var err error
+	m.container, err = m.createV3ioClient(session, "", token)
```

Comment:
Avoid creating per request to avoid go routine leak.

Reply:
i am creating the fasthttp client and v3io-go context only once

Reply:
Moved the comment to the relevant line. fasthttp isn't the one running go routines in the background. v3io-go does this in the context.